[iOS/Critical] Fix ListView memory leaks

adrianknight89 opened the pull request:
### Description of Change ###

iOS `ListView` is leaking for various reasons. Although the bug description mentions issues with `RecycleElement`, this applies to either caching strategy. Since this seems to be a serious issue, can this PR be given priority?

See comments for details.

P.S. It was mentioned that the same issue occurs on Android, but I wasn't able to reproduce it on AppCompat (at least not with the sample code in this PR).

EDIT: Also included the fix for another bug concerning context actions not being properly disposed.

### Bugs Fixed ###

- https://bugzilla.xamarin.com/show_bug.cgi?id=43941
- https://bugzilla.xamarin.com/show_bug.cgi?id=32206
- https://bugzilla.xamarin.com/show_bug.cgi?id=27194 (possibly fixes this too. Download links seem to be down.)

### PR Checklist ###

- [x] Has tests (if omitted, state reason in description)
- [x] Rebased on top of master at time of PR
- [x] Changes adhere to coding standard
- [x] Consolidate commits as makes sense

adrianknight89 commented on Xamarin.Forms.Platform.iOS/Cells/CellTableViewCell.cs:
```diff
@@ -95,5 +95,21 @@ internal static UITableViewCell GetNativeCell(UITableView tableView, Cell cell,
+			_disposed = true;
+
+			base.Dispose(disposing);
+		}
```
One of the issues seems to be that `CellTableViewCell` is hanging onto `_cell`. This should release the reference.

adrianknight89 commented on Xamarin.Forms.Platform.iOS/Renderers/ListViewRenderer.cs:
```diff
@@ -91,24 +92,25 @@ public override void LayoutSubviews()
+					foreach (UIView child in view.Subviews)
+						viewsToLookAt.Push(child);
+
+					view.Dispose();
```
As this loop executes, eventually we hit `CellTableViewCell`. Calling `Dispose` on it should properly release the reference to its `_cell`.

The original code was calling `Dispose` only on `ViewCellRenderer.ViewTableCell` instances. At the moment, I'm not sure if disposing every subview is the correct approach, so this is up for discussion. Please note that `TableViewRenderer` is doing the same thing, so it might need an update as well.

adrianknight89 replied:
EDIT: I made changes so that subviews are visited recursively, removed from their superviews, and disposed before their parents/superviews can be disposed.

adrianknight89 commented on Xamarin.Forms.Platform.iOS/Renderers/ListViewRenderer.cs:
```diff
@@ -118,27 +120,28 @@ protected override void Dispose(bool disposing)
+				{
+					_dataSource.Dispose();
+					_dataSource = null;
+				}
```
The data source needs to be disposed as well. See what it's doing below.

adrianknight89 commented on Xamarin.Forms.Platform.iOS/Renderers/ListViewRenderer.cs:
```diff
@@ -650,21 +656,37 @@ internal nfloat CalculateHeightForCell(UITableView tableView, Cell cell)
+				_disposed = true;
+
+				base.Dispose(disposing);
+			}
```
Release the reference to `_prototype` but do not dispose it.

adrianknight89 commented on Xamarin.Forms.Platform.iOS/Renderers/ListViewRenderer.cs:
```diff
@@ -984,6 +1006,29 @@ void UpdateShortNameListener()
+				_disposed = true;
+
+				base.Dispose(disposing);
+			}
```
I did not see `OnItemSelected` being unregistered anywhere. References to `ListView` and `FormsUITableViewController` must be nullified. I think what is happening is when the constructor is called, we are passing the references by value which creates a copy for each. You could test this particular change by having a list view with no items. If these references are not nullified, you end up with a leak still.

I'm not sure if setting `_templateToId` to null was required since it's a `Dictionary`, but the keys are data templates. It seemed safer to null it out as well.

adrianknight89 commented on Xamarin.Forms.Platform.iOS/Renderers/ListViewRenderer.cs:
```diff
@@ -1085,15 +1131,25 @@ public override void ViewWillAppear(bool animated)
+
+			_disposed = true;
+
+			base.Dispose(disposing);
```
Same thing with `_list`

hartez commented:
At first glance this looks pretty good; I'll have to spend some time on Monday looking more closely at the [subview disposal stuff you mentioned](https://github.com/xamarin/Xamarin.Forms/pull/524#discussion_r87699000). 

Could you add some instructions to the test so when the case is being manually verified the tester knows what to look for?

adrianknight89 commented:
@hartez Done.

adrianknight89 commented:
@hartez I included the fix for another bug as well. Also added a disposed check to `ViewTableCell` since it was being multi-disposed.

hartez commented:
@adrianknight89 Current version is failing to build; looks like there's a conflict between your `CustomPage` class and another UI test.

adrianknight89 commented:
@hartez Can you try now?

adrianknight89 commented:
I think some of these references can be converted to `WeakReference`.

EmmanVazz commented:
I have also been able to reproduce these leaks and this seem like a good fix. I think these leaks also affect CarouselView as I found a similar leak while using it. I would think it requires a similar fix to this.

hartez commented on Xamarin.Forms.Controls.Issues/Xamarin.Forms.Controls.Issues.Shared/Bugzilla32206.cs:
```diff
@@ -0,0 +1,78 @@
+		}
+	}
+
+	internal class CustomPageY : ContentPage
```
Instead of `CustomPageY` and `CustomPageX`, could you give these names which are less likely to accidentally conflict with other test cases in the future? Using the issue number in the name should be sufficient; something like `_43941CustomPage` or `CustomPage43941`.

hartez commented on Xamarin.Forms.Controls.Issues/Xamarin.Forms.Controls.Issues.Shared/Bugzilla32206.cs:
```diff
@@ -0,0 +1,78 @@
+			});
+		}
+	}
+
```
You'll need to add the `[Preserve(AllMembers = true)]` attribute to these supplemental classes in your test cases; otherwise, the [linker](https://developer.xamarin.com/guides/ios/advanced_topics/linker/#Preserving_Code) will remove them and the automated UI tests will fail.

hartez commented on Xamarin.Forms.Controls.Issues/Xamarin.Forms.Controls.Issues.Shared/Bugzilla43941.cs:
```diff
@@ -0,0 +1,72 @@
+﻿using System;
```
Both this and the test case for 32206 could be automated. Take a look at the test for [44166](https://github.com/xamarin/Xamarin.Forms/blob/122f0e3e6060711d6c3b2b8523236b8a057434df/Xamarin.Forms.Controls.Issues/Xamarin.Forms.Controls.Issues.Shared/Bugzilla44166.cs) for an example.

hartez commented on Xamarin.Forms.Controls.Issues/Xamarin.Forms.Controls.Issues.Shared/Bugzilla32206.cs:
```diff
@@ -0,0 +1,78 @@
+using NUnit.Framework;
+#endif
+
+namespace Xamarin.Forms.Controls
```
Namespace should be `Xamarin.Forms.Controls.Issues` (trying to get that to be more consistent).

hartez commented on Xamarin.Forms.Controls.Issues/Xamarin.Forms.Controls.Issues.Shared/Bugzilla43941.cs:
```diff
@@ -0,0 +1,72 @@
+using NUnit.Framework;
+#endif
+
+namespace Xamarin.Forms.Controls
```
Namespace should be `Xamarin.Forms.Controls.Issues`.

adrianknight89 commented:
@hartez I addressed the issues except for the UI tests. When I try to compile `Xamarin.Forms.ControlGallery.iOS` in release mode in XS with the latest alpha of Xamarin, I'm getting a linker issue:

```
/Users/xyz/Projects/adrianknight89/Xamarin.Forms/Xamarin.Forms.ControlGallery.iOS/Xamarin.Forms.ControlGallery.iOS.csproj (Build) ->
/Library/Frameworks/Mono.framework/External/xbuild/Xamarin/iOS/Xamarin.iOS.Common.targets (_CompileToNative target) ->

	MTOUCH: error MT2001: Could not link assemblies. Reason: Value cannot be null.
```

hartez commented:
@adrianknight89 It's a [known issue in the alpha](https://bugzilla.xamarin.com/show_bug.cgi?id=46510); if you switch to the stable channel you should be able to build.

It looks like they've already got a fix for it, so I'm guessing you'll be able to build with the next alpha update as well.

adrianknight89 commented:
@hartez I have added a UI test for `32206`. It navigates back and forth on the stack 10 times and hits `GC` button to collect and finalize garbage data. The test should be successful if the counter is zero. However, it seems static variables I added for tracking state seem to be null because the parent page no longer exists. If you run the test now, it will crash because `_label` (tried counter before as well) is null. Would you mind telling me why this is the case? Before I proceed to the other test, I'd like to figure this one out first.

Also came across another bug where, if you push and pop the page rapidly, eventually a page will start leaking. This could be a problem with `NavigationPage`. It's something to keep in mind for now.

rmarinho commented:
@adrianknight89 you can also disable the linker if you want to workaround that error for now.

adrianknight89 commented:
@rmarinho I worked around the first issue by going back to the stable channel. You mean the static variable issue is also a linker issue?

rmarinho commented:
Nop i was talking if you want to build in alpha and go around that linker issue

hartez commented:
@adrianknight89 I think making the Label static is causing your problems. Change it to

`public Label _label;`

And instead of accessing it from the landing page like this:
```
if (LandingPage32206._label.Text != "Counter: 0")
    Assert.Fail("Test page is leaking.");
```

Just let the UI test framework look for the correct value:

```
RunningApp.WaitForElement(q => q.Marked("Counter: 0"));
```

adrianknight89 commented:
@hartez Your suggestion seems to work; however, I still believe `44166` is broken because it's querying static variables. 

I added the other UI test as well. Let me know if any other changes are needed.

adrianknight89 commented:
Found out what the issue was with regards to the extra renderer being created. It happens when uneven rows are used to calculate an estimated row height. Added code to properly dispose `_prototype`.

adrianknight89 commented:
@hartez There seems to be an issue with the way `_prototype` is disposed when its element is a layout. Disposing the renderer is not enough to dispose child renderers.

`HasUnevenRows=true`

```
public CustomViewCell : ViewCell
{
    public CustomViewCell()
    {
         View = new DatePicker();
         //View = new StackLayout
         // {
         //        Children = {new DatePicker()}
         // }
    }
}
```

`CalculateHeightForCell` is clearing renderers out of elements like so:

```
target.ClearValue(Platform.RendererProperty);
foreach (var descendant in target.Descendants())
	descendant.ClearValue(Platform.RendererProperty);
```

This doesn't seem to dispose renderers.

adrianknight89 commented:
Disposing child renderers while clearing `RendererProperty` in `CalculateHeightForCell` seems to fix the above issue.

jonathanpeppers commented:
@adrianknight89 we are using this fix in our fork of Xamarin.Forms. Thanks for it, you have so many PRs.

There was one change I had to make that was causing NRE's in our app: https://github.com/Hitcents/Xamarin.Forms/commit/8f2bed61f86eb9f15696a4f74cab4f029c96e618

adrianknight89 commented:
@jonathanpeppers Thanks for your input. Was the NRE caused by my changes? I'm not sure if we should always expect Element to be non-null by the time we're accessing the platform object.

jonathanpeppers commented:
So we put your changes on top of our fork by itself. We were specifically trying to fix a memory leak with `ListView` (which is fixed, thanks).

I guess because we have lots of headers, it could cause the NRE. It is also possible we have a custom renderer as a header.

rmarinho commented:
can you add that check @adrianknight89 ?

adrianknight89 commented:
Done.
